File: Carts/Controllers/Scan and Search Controller/CAScanningSession.m
//
//  CAScanningSession.m
//  GroceryList
//
//  Created by Hakon Hanesand on 3/9/15.

#import <ReactiveCocoa/ReactiveCocoa.h>

#import "CAScanningSession.h"
#import "CABarcodeObject.h"
#import "CABarcode.h"
#import "UIImage+CAImage.h"

@import AVFoundation;

@interface CAScanningSession()
@property (nonatomic) AVCaptureDevice *captureDevice;
@property (nonatomic) AVCaptureDeviceInput *videoInput;
@property (nonatomic) AVCaptureMetadataOutput *metadataOutput;
@property (nonatomic) AVCaptureVideoDataOutput *videoOutput;

@property (nonatomic) AVCaptureStillImageOutput *imageCapture;

@property (nonatomic) AVCaptureConnection *connection;
@property (nonatomic) BOOL paused;

@property (nonatomic) UIImage *recievedImage;
@end

@implementation CAScanningSession

+ (CAScanningSession *)session {
    return [[CAScanningSession alloc] init];
}

- (instancetype)init {
    if (self = [super init]) {
        [self initializeCaptureSession];
        [self setupCaptureSignal];
        [self.metadataOutput setMetadataObjectTypes:self.metadataOutput.availableMetadataObjectTypes];
    }
    
    return self;
}

- (void)setupCaptureSignal {
    //captures barcodes from the camera and maps them to CABarcodes
    
    RACSignal *barcodeSelectorSignal = [self rac_signalForSelector:@selector(captureOutput:didOutputMetadataObjects:fromConnection:) fromProtocol:@protocol(AVCaptureMetadataOutputObjectsDelegate)];
    
    self.barcodeSignal = [[[[[barcodeSelectorSignal filter:^BOOL(RACTuple *value) {
        return !self.paused && [[value.second firstObject] isKindOfClass:[AVMetadataMachineReadableCodeObject class]]; //optimization
    }] bufferWithTime:0.5 onScheduler:[RACScheduler scheduler]] map:^CABarcode *(RACTuple *buffer) {
        RACTuple *firstSelectorTuple = buffer.first;
        NSArray *barcodeMetadataObjects = firstSelectorTuple.second;
        return [CABarcode barcodeWithMetadataObject:[barcodeMetadataObjects firstObject]];
    }] logNext] filter:^BOOL(id value) {
        return !self.paused; //double check that we are not paused
    }];
}

- (void)pause {
    self.paused = YES;
}

- (void)resume {
    self.paused = NO;
}

- (void)start {
    if (!self.captureSession.running) {
        dispatch_async(dispatch_get_CAobal_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
            [self.captureSession startRunning];
        });
    }
    
}

- (void)stop {
    if (self.captureSession.running) {
        dispatch_async(dispatch_get_CAobal_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
            [self.captureSession stopRunning];
        });
    }
}

- (void)initializeCaptureSession {
    if (self.captureSession) {
        return;
    }
    
    self.captureSession = [AVCaptureSession new];
    self.captureSession.sessionPreset = AVCaptureSessionPresetPhoto;
    
    self.captureDevice = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    
    if (!self.captureDevice) {
        NSLog(@"No video camera on this phone!");
        return;
    }
    
    NSError *error;
    
    self.videoInput = [[AVCaptureDeviceInput alloc] initWithDevice:self.captureDevice error:&error];
    
    if ([self.captureSession canAddInput:self.videoInput]) {
        [self.captureSession addInput:self.videoInput];
    } else {
        NSLog(@"Error %@", error);
    }
    
    AVCaptureVideoDataOutput *videoOutput = [[AVCaptureVideoDataOutput alloc] init];
    
    if ([self.captureSession canAddOutput:videoOutput]) {
        [self.captureSession addOutput:videoOutput];
    }
    
    videoOutput.videoSettings = [NSDictionary dictionaryWithObject:[NSNumber numberWithInt:kCVPixelFormatType_32BGRA] forKey:(id)kCVPixelBufferPixelFormatTypeKey];
    
    [videoOutput setSampleBufferDelegate:self queue:dispatch_queue_create("dispatch", 0)];
    
    self.metadataOutput = [[AVCaptureMetadataOutput alloc] init];
    [self.metadataOutput setMetadataObjectsDelegate:self queue:dispatch_get_main_queue()];
    
    if ([self.captureSession canAddOutput:self.metadataOutput]) {
        [self.captureSession addOutput:self.metadataOutput];
    }
}

- (RACSignal *)captureImageFromVideoOutput {
    self.connection.enabled = YES; //start the output of frames and after recieving one
    
    //observe the recievedImage property for changes
    return [[RACObserve(self, recievedImage) skip:1] take:1];
}

//saves the connection to a property (so it can be resumed later), then disables the connection
- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection {
    self.connection = connection;
    connection.enabled = NO;
    
    //set the recieved image property so observers can fire signals
    self.recievedImage = [UIImage imageFromSampleBuffer:sampleBuffer];
}

@end

File: Carts/Controllers/Scan and Search Controller/CAScanningSession.swift
//
//  CAScanningSession.swift
//  Carts
//
//  Created by Hakon Hanesand on 4/26/15.
//  Copyright (c) 2015 Hakon Hanesand. All rights reserved.
//

import UIKit
import AVFoundation

extension Array {
    
    private func all(matching: Element -> Bool) -> Bool {
        return self.map(matching).reduce(true) { $0 && $1 }
    }
}

class CAScanningSession: NSObject, AVCaptureMetadataOutputObjectsDelegate, AVCaptureVideoDataOutputSampleBufferDelegate {
//    var captureDevice: AVCaptureDevice?
    
    var connection: AVCaptureConnection?
    var paused: Bool
    
    var recievedImage: UIImage?
    
    var barcodeSignal: SignalProducer<String, NSError>
    
    var captureSession: AVCaptureSession!
    
    class func session() -> CAScanningSession {
        return CAScanningSession()
    }
    
    override required init() {
        paused = true
        
        super.init()
        
        setupBarcodeSignal()
        initializeCaptureSession()
    }
    
    func setupBarcodeSignal() {
        let barcodeSelectorSignal = self.rac_signalForSelector(#selector(AVCaptureMetadataOutputObjectsDelegate.captureOutput(_:didOutputMetadataObjects:fromConnection:)), fromProtocol: "AVCaptureMetadataOutputObjectsDelegate" as! Protocol).toSignalProducer().map { return $0 as! RACTuple }
        
        barcodeSignal = barcodeSelectorSignal.filter {
            let metadataObjects = $0.second as! [AVMetadataObject]
            let barcodeMetadataObjects = metadataObjects.filter { $0.isKindOfClass(AVMetadataMachineReadableCodeObject) }
            return !self.paused && barcodeMetadataObjects.count > 0
        }.logEvents().map { test in
            // TODO - Return barcode created from AVMetadataMachineReadableCodeObject
            return ""
        }
    }
    
    func start() {
        if self.captureSession!.running {
            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), {
                self.captureSession!.startRunning()
            })
        }
    }
    
    func stop() {
        if self.captureSession!.running {
            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), {
                self.captureSession!.stopRunning()
            })
        }
    }
    
    private func initializeCaptureSession() {        
        self.captureSession = AVCaptureSession()
        self.captureSession.sessionPreset = AVCaptureSessionPresetPhoto
    
        let captureDevice = AVCaptureDevice.defaultDeviceWithMediaType(AVMediaTypeVideo)
        
        if let videoInput = try? AVCaptureDeviceInput(device: captureDevice) {
            self.captureSession.addInput(videoInput)
            let videoOutput = AVCaptureVideoDataOutput()
            self.captureSession.addOutput(videoOutput)
            
            videoOutput.videoSettings = [Int(kCVPixelFormatType_32BGRA) : kCVPixelBufferPixelFormatTypeKey]
            videoOutput.setSampleBufferDelegate(self, queue: dispatch_queue_create("dispatch"))
            
            let metadataOutput = AVCaptureMetadataOutput()
            metadataOutput.setMetadataObjectsDelegate(self, queue: dispatch_get_main_queue())
            self.captureSession?.addOutput(metadataOutput)
            
            metadataOutput.metadataObjectTypes = metadataOutput.availableMetadataObjectTypes
        } else {
            fatalError("Failed to initialize video.")
        }
        
    }
    
    func captureImageFromVideoOutput() {
//        self.connection.enabled = YES; //start the output of frames and after recieving one
//        
//        //observe the recievedImage property for changes
//        return [[RACObserve(self, recievedImage) skip:1] take:1];
    }
    
    func captureOutput(captureOutput: AVCaptureOutput!, didDropSampleBuffer sampleBuffer: CMSampleBuffer!, fromConnection connection: AVCaptureConnection!) {
//        self.connection = connection;
//        connection.enabled = NO;
//        
//        //set the recieved image property so observers can fire signals
//        self.recievedImage = [UIImage imageFromSampleBuffer:sampleBuffer];
    }
}
